lib/autoupdate/agent.go

```diff
@@ -65,30 +86,204 @@ type AgentUpdateStatus struct {
+}
+
+// Enable enables agent updates and attempts an initial update.
+// If the initial update succeeds, auto-updates are enabled and the configuration is persisted.
+// Otherwise, the auto-updates configuration is not changed.
+// This function is idempotent.
+func (u *AgentUpdater) Enable(ctx context.Context, userCfg AgentUserConfig) error {
+	// Read configuration from updates.yaml and override any new values passed as flags.
+	cfg, err := u.readConfig(u.ConfigPath)
+	if err != nil {
+		return trace.Errorf("failed to read update.yaml: %w", err)
+	}
+	if userCfg.Proxy != "" {
+		cfg.Spec.Proxy = userCfg.Proxy
+	}
+	if userCfg.Group != "" {
+		cfg.Spec.Group = userCfg.Group
+	}
+	if userCfg.URLTemplate != "" {
+		cfg.Spec.URLTemplate = userCfg.URLTemplate
+	}
+	cfg.Spec.Enabled = true
+	if err := validateUpdatesSpec(&cfg.Spec); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Lookup target version from the proxy.
+	addr, err := libutils.ParseAddr(cfg.Spec.Proxy)
+	if err != nil {
+		return trace.Errorf("failed to parse proxy server address: %w", err)
+	}
+
+	desiredVersion := userCfg.ForceVersion
+	if desiredVersion == "" {
+		resp, err := webclient.Find(&webclient.Config{
+			Context:   ctx,
+			ProxyAddr: addr.Addr,
+			Insecure:  u.InsecureSkipVerify,
+			Timeout:   30 * time.Second,
+			//Group:     cfg.Spec.Group, // TODO(sclevine): add web API
+			Pool: u.Pool,
+		})
+		if err != nil {
+			return trace.Errorf("failed to request version from proxy: %w", err)
+		}
+		desiredVersion, _ = "16.3.0", resp // TODO(sclevine): add web API
+		//desiredVersion := resp.AutoUpdate.AgentVersion
+	}
+
+	if desiredVersion == "" {
+		return trace.Errorf("agent version not available from Teleport cluster")
+	}
+	// If the active version and target don't match, kick off upgrade.
+	if cfg.Status.ActiveVersion != desiredVersion {
+		template := cfg.Spec.URLTemplate
+		if template == "" {
+			template = cdnURITemplate
+		}
+		// Create /var/lib/teleport/versions/X.Y.Z if it does not exist.
+		err = u.Installer.Install(ctx, desiredVersion, template)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		cfg.Status.ActiveVersion = desiredVersion
+		u.Log.InfoContext(ctx, "Target version successfully installed.", "version", desiredVersion)
+	} else {
+		u.Log.InfoContext(ctx, "Target version already installed.", "version", desiredVersion)
+	}
+
+	// Always write the configuration file if enable succeeds.
+	if err := u.writeConfig(u.ConfigPath, cfg); err != nil {
+		return trace.Errorf("failed to write update.yaml: %w", err)
+	}
+	u.Log.InfoContext(ctx, "Configuration updated.")
+	return nil
+}
+
+func validateUpdatesSpec(spec *AgentUpdateSpec) error {
+	if spec.URLTemplate != "" &&
+		!strings.HasPrefix(strings.ToLower(spec.URLTemplate), "https://") {
+		return trace.Errorf("Teleport download URL must use TLS (https://)")
+	}
+
+	if spec.Proxy == "" {
+		return trace.Errorf("Teleport proxy URL must be specified with --proxy or present in updates.yaml")
```

Comment:
`update.yaml` or we might use `agentUpdateConfigName` constant for formatting the error message